### ISSUE TITLE
Move more work to background isolate?

### ISSUE THREAD

**Issue by dnfield** [simolus3/moor#2279]
I've been looking at a large application's performance using Drift.

One thing I'm seeing in profiling is that significant time can be spent on the UI isolate for calls to `DatabaseConnectionUser.batch` for larger entities on older phones.

This seems like something that should be possible to move onto the background isolate, especially when the background isolate is in the same isolate group.

It may be possible for consumers of this library to handle this, but it's not entirely clear to me what the best way for them to do that is - and it may be surprising to consumers of this library that they have to do it if they're doing their SQL work on a background isolate anyway.

Do you have any thoughts on making this work more cleanly within the library?

**Comment by simolus3** [comment]
Thanks for the suggestion!

> It may be possible for consumers of this library to handle this, but it's not entirely clear to me what the best way for them to do that is

Generally, it's possible to share a logical drift database instance between isolates (once you have a `DriftIsolate`, it can be sent across ports and multiple "client" isolates can connect to it). So the whole task of inserting bulk data in a batch could be moved into a `compute()`-like worker that shares the logical database with the UI.

> it may be surprising to consumers of this library that they have to do it if they're doing their SQL work on a background isolate anyway

That's a fair point, I think drift should offload more work to background isolates where that's possible. At the moment, drift uses isolates at a fairly low level (executing SQL), which has the advantage of being very general (like supporting a similar mechanism based on web workers) at the cost of not being able to send arbitrary Dart objects.

Intuitively, I can see two APIs to approach this:

1. Explicitly: Drift could have a `computeWithDatabase` function that's similar to Flutter's `compute`, except that the callback also receives a drift database instance in addition to the argument:
```dart
await computeWithIsolate((arg, db) async {
  // Runs in a background isolate, db is logically the same database as foregroundDb
}, argToWorker, foregroundDb);
````
2. Implicitly: When a drift database is connected to an existing isolate, some work like constructing SQL could implicitly happen on the background isolate, without requiring any API change.

Would something like the first approach be feasible in your use-case? A downside is that it can only be used for things that can be offloaded to `compute`, so it doesn't really work if constructing the batch depends on shared mutable state.
In your profiling analysis, is most of the time spent in `DatabaseConnectionUser.batch` in application code or in other batch APIs responsible for constructing SQL? `batch` invokes a callback to collect SQL statements, so if a lot of time is generally spent in the callback that would explain `batch()` standing out. If most is spent in other drift methods, an implicit approach to offload this work may be beneficial.

**Comment by dnfield** [comment]
I'm hoping (perhaps in vain) that some change could be made in the package so more of the work to build the query/DAO objects could be done on the same isolate as the actual query is executed on.

If it has to be new API that's probably ok. The only problem will be to take care to make sure that the argument(s) being passed are safe to send over a port.

**Comment by dnfield** [comment]
So, for example, if I wrap https://github.com/simolus3/drift/blob/develop/drift/test/isolate_test.dart#L309 in `Isolate.run`, I get this:

```
 dart test test/isolate_test.dart
00:01 +8 -1: in same isolate with explicit serialization can run deletes, updates and batches [E]                                                                                                                                                                                                                    
  Invalid argument(s): Illegal argument in isolate message: (object is a SuspendState)
  dart:isolate                   Isolate.run
  test/isolate_test.dart 255:19  _runTests.<fn>
```

I'm not even sure where that object is coming from though. Ideally Dart would also be giving me a better exception, which I'll be fliing a bug for upstream.

**Comment by simolus3** [comment]
Yeah, a whole `DatabaseConnectionUser` can't be shared across isolates. My point is that you can send the stateful part of a connection across isolates though:

```dart
import 'dart:isolate';

import 'package:drift/drift.dart';
import 'package:drift/isolate.dart';
import 'package:drift/native.dart';
import 'package:uuid/uuid.dart';

part 'isolate.g.dart';

class TextRows extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get content => text()();
}

@DriftDatabase(tables: [TextRows])
class MyDatabase extends _$MyDatabase {
  MyDatabase(super.connection) : super.connect();

  @override
  int get schemaVersion => 1;
}

void main() async {
  final driftServer = await DriftIsolate.spawn(
      () => DatabaseConnection(NativeDatabase.memory()));

  final databaseForMainIsolate = MyDatabase(await driftServer.connect());
  databaseForMainIsolate
      .select(databaseForMainIsolate.textRows)
      .watch()
      .listen(print);

  // We can now offload some work onto a remote isolate
  await Isolate.run(() async {
    // Despite being separate instances, the two databases on the different
    // isolates share their state.
    final backgroundDatabase = MyDatabase(await driftServer.connect());

    await backgroundDatabase.batch((batch) {
      for (var i = 0; i < 1000; i++) {
        batch.insert(backgroundDatabase.textRows,
            TextRowsCompanion.insert(content: Uuid().v4()));
      }
    });
  });
}
```

I think the main goal here should be to reduce the amount of code necessary to get this running.

**Comment by simolus3** [comment]
> that some change could be made in the package so more of the work to build the query/DAO objects could be done on the same isolate as the actual query is executed on.

At the moment, `batch` synchronously throws an exception when attempting to insert invalid data (e.g. a row that doesn't have values for non-nullable columns). So we probably can't change the existing behavior without making it a breaking change.

Due to [type converters](https://drift.simonbinder.eu/docs/advanced-features/type_converters/), the process of building SQL from Dart types may also involve running user code, and it could mean that the starting object can't be sent over isolates to begin with. So unfortunately it seems like a change in drift itself to magically make this happen without any user control is impossible.

-----

To make the previous example much easier to write, drift could provide an API that let's you do this:

```dart
final database = MyDatabase(/* can be constructed without special consideration to isolates */):

final DatabaseConnection exposedConnection = database.independentConnection(); // proposed API
await Isolate.run(() async {
  final database = MyDatabase(exposedConnection);
});
```

**Comment by dnfield** [comment]
I think this is promising, but I'm not deep enough into the application I was testing to know how big of a lift it would be to refactor to use that api.

**Comment by simolus3** [comment]
I have added two helpers for this: https://github.com/simolus3/drift/blob/28f31d183ee50f4b327ac493da7689600d797e7a/drift/lib/isolate.dart#L189-L322

The `serializableConnection()` method returns a `DriftIsolate` that can be sent across ports, so you can connect to it in a worker. `computeWithDatabase` is essentially a wrapper around `Isolate.run` that also passes a copy of the database class to the callback, which could be useful to offload some tasks like an entire `batch`.

This has already been possible before by manually spawning a `DriftIsolate` and then having both the main isolate and temporary workers connect to it, but those utilities should make it a lot easier.

> but I'm not deep enough into the application I was testing to know how big of a lift it would be to refactor to use that api.

I hope the `computeWithDatabase` api is fairly easy to adopt, especially when only used for database operations. If the general pattern around the expensive batch looks like this:

```dart
final rows = await loadRowsToInsert();
await database.batch((batch) => batch.insertAll(someTable, rows));
```

The version of doing the SQL-generation on a new isolate would now be:

```dart
final rows = await loadRowsToInsert();
await database.computeWithDatabase(
  connect: DatabaseClass.new,
  computation: (db) async {
    await db.batch((batch) => batch.insertAll(someTable, rows));
  },
);
```

There's still the cost of sending `rows` to the new isolate of course, but that's probably the easiest optimization if `loadRowsToInsert()` can't be moved onto a new isolate.

**Comment by 3rpse-ai** [comment]
Hi, 
I have tried to use the `computeWithDatabase(...)` method, yet I could not get it to work. In particular I could not use the connect argument like you did in the example (or the doc comments) as `DatabaseClass.new` constructor does not expect any arguments. I am using drift 2.5.0. Do I need some additional setup to make this work?

**Comment by simolus3** [comment]
Hi, yes you need a database constructor like this:

```dart
@DriftDatabase(...)
class MyDatabase extends _$MyDatabase {
  MyDatabase(): super(_createDatabaseConnection()); // your existing constructor probably looks like this
  MyDatabase.custom(QueryExecutor e): super(e); // You can add a new one like this
}
```

Then, you'd use `MyDatabase.custom` instead of `MyDatabase.new`. This came up multiple times, I'll improve the documentation to make that clearer.

**Comment by 3rpse-ai** [comment]
Thanks a lot for your quick reply! However now after adding a suiting constructor I get this error after running further queries on the initial db:
```
StateError (Bad state: Can't re-open a database after closing it. Please create a new database connection and open that instead.)
```
Basically what I am doing is to fetch a huge list of ids in the `computeWithDatabase(...)` method. Then I want to access single objects via the ids. As the fetching of the ids caused the app to slightly stutter I wanted to move it to an isolate. 

So do I need to create now a new db instance whenever I have used `computeWithDatabase(...)` ?


Edit:
I managed to achieve the same by using the `NativeDatabase.createInBackground(file)` constructor (which I missed before unfortunately). This solved my problem like a charm. However I would still be interested about how to solve my issue above if you have some time :)

**Comment by simolus3** [comment]
Thanks for the follow-up! That's a (rather embarrassing)  bug in drift that I've fixed in 94c915504deee9c48f07b84fcf407e53736819b2.

Using `createInBackground` with `computeWithDatabase` will work as well. It's just the normal  `NativeDatabase` that unfortunately gets closed by `computeWithDatabase`. Regardless of whether `computeWithDatabase` is used or not, using `createInBackground` is a good idea though :)

**Comment by 3rpse-ai** [comment]
Thanks again for the quick reply & fix! Will check it out again once the fix gets released. :)
